client/rpc-servers/src/middleware.rs

```diff
@@ -18,46 +18,69 @@
 
 //! Middleware for RPC requests.
 
+use std::time::Instant;
+
 use jsonrpc_core::{
 	Middleware as RequestMiddleware, Metadata,
-	Request, Response, FutureResponse, FutureOutput
-};
-use prometheus_endpoint::{
-	Registry, CounterVec, PrometheusError,
-	Opts, register, U64
+	FutureResponse, FutureOutput
 };
+use prometheus_endpoint::{GaugeVec, HistogramOpts, HistogramVec, Opts, PrometheusError, Registry, U64, register};
 
 use futures::{future::Either, Future};
 
 /// Metrics for RPC middleware
 #[derive(Debug, Clone)]
 pub struct RpcMetrics {
-	rpc_calls: Option<CounterVec<U64>>,
+	rpc_calls: HistogramVec,
+	started: GaugeVec<U64>,
+	finished: GaugeVec<U64>
 }
 
 impl RpcMetrics {
 	/// Create an instance of metrics
-	pub fn new(metrics_registry: Option<&Registry>) -> Result<Self, PrometheusError> {
-		Ok(Self {
-			rpc_calls: metrics_registry.map(|r|
-				register(
-					CounterVec::new(
-						Opts::new(
-							"rpc_calls_total",
-							"Number of rpc calls received",
+	pub fn new(metrics_registry: Option<&Registry>) -> Result<Option<Self>, PrometheusError> {
+		if let Some(r) = metrics_registry {
+			Ok(Some(Self {
+				rpc_calls: register(
+					HistogramVec::new(
+						HistogramOpts::new(
+							"rpc_calls_time",
+							"Total time [ms] of processed RPC calls",
 						),
-						&["protocol"]
+						&["protocol", "method"]
 					)?,
 					r,
-				)
-			).transpose()?,
-		})
+				)?,
+				started: register(
+					GaugeVec::new(
+						Opts::new(
+							"rpc_calls_started",
+							"Number of received RPC calls (unique un-batched requests)"
+						),
+						&["protocol", "method"]
+					)?,
+					r
+				)?,
+				finished: register(
+					GaugeVec::new(
+						Opts::new(
+							"rpc_calls_finished",
```

Comment:
Are these calls separated by transport type? I.e. http/ws?

Reply:
Yes, we have `protocol` label.